src/components/utils/src/timer.cc

```diff
@@ -44,175 +44,176 @@ CREATE_LOGGERPTR_GLOBAL(logger_, "Utils")
-  Stop();
-  sync_primitives::AutoLock auto_lock(task_lock_);
-  DCHECK(task_);
-  delete task_;
+
+  DCHECK(thread_);
+  thread_->join();
+  DeleteThread(thread_);
 }
 
 void timer::Timer::Start(const Milliseconds timeout, const bool repeatable) {
   LOG4CXX_AUTO_TRACE(logger_);
   sync_primitives::AutoLock auto_lock(lock_);
-  SetTimeoutUnsafe(timeout);
-  repeatable_ = repeatable;
-  if (is_running_) {
-    const bool stop_result = StopUnsafe();
-    DCHECK_OR_RETURN_VOID(stop_result);
+
+  LOG4CXX_DEBUG(logger_,
+                "Prepare start timer for" << timeout << "ms, repetable = "
+                                          << (repeatable ? "true" : "false"));
+
+  if (repeatable) {
+    delegate_.make_repeatable();
+  }
+
+  UpdateTimeOut(timeout);
+
+  if (thread_->is_running()) {
+    LOG4CXX_INFO(logger_, "Restart timer in thread " << name_);
+    delegate_.ShouldBeRestarted();
+  } else {
+    LOG4CXX_INFO(logger_, "Start new timer in thread " << name_);
+    thread_->start();
   }
-  StartUnsafe();
+
+  delegate_.WaitUntilStart();
 }
 
 void timer::Timer::Stop() {
   LOG4CXX_AUTO_TRACE(logger_);
   sync_primitives::AutoLock auto_lock(lock_);
-  repeatable_ = false;
-  if (is_running_) {
-    const bool stop_result = StopUnsafe();
-    DCHECK(stop_result);
+  DCHECK(thread_);
+  LOG4CXX_DEBUG(logger_, "Stopping timer  " << name_);
+  if (pthread_equal(pthread_self(), thread_->thread_handle())) {
+    // Thread can't stop itself , so it will suspend
+    LOG4CXX_DEBUG(logger_, "Suspend timer " << name_ << " after next loop");
+    delegate_.ShouldBeStopped();
+  } else {
+    thread_->join();
   }
 }
 
 bool timer::Timer::IsRunning() const {
   sync_primitives::AutoLock auto_lock(lock_);
-  return is_running_;
+  return (thread_->is_running() && !delegate_.IsGoingToStop());
 }
 
-void timer::Timer::OnTimeout() {
+void timer::Timer::UpdateTimeOut(const Milliseconds timeout_milliseconds) {
+  // There would be no way to stop thread if timeout in lopper will be 0
+  uint32_t timeout = (timeout_milliseconds > 0u) ? timeout_milliseconds : 1u;
+
+  LOG4CXX_DEBUG(logger_,
+                "Set new timeout " << timeout << "ms for timer " << name_);
+  delegate_.SetTimeOut(timeout);
+}
+
+void timer::Timer::OnTimeout() const {
   LOG4CXX_AUTO_TRACE(logger_);
   LOG4CXX_DEBUG(logger_,
                 "Timer has finished counting. Timeout(ms): "
-                    << static_cast<uint32_t>(timeout_ms_));
-  {
-    // Task locked by own lock because from this task in callback we can
-    // call Stop of this timer and get DeadLock
-    sync_primitives::AutoLock auto_lock(task_lock_);
-    DCHECK(task_);
-    task_->run();
-  }
-  sync_primitives::AutoLock auto_lock(lock_);
-  if (is_running_) {
-    const bool stop_result = StopUnsafe();
-    DCHECK_OR_RETURN_VOID(stop_result);
-  }
-  if (repeatable_) {
-    StartUnsafe();
-  }
+                    << static_cast<uint32_t>(delegate_.get_timeout()));
+
+  DCHECK(task_.get());
```

Comment:
1. `DCHECK_OR_RETUNR`
2. It looks `get` is redundant - we can just use `DCHECK(tesk_)` indirectly, can't we?